federation-js/src/service/printCoreSchema.ts

```diff
@@ -0,0 +1,581 @@
+ *
+ */
+// Core change: we need service and url information for the join__Graph enum
+export function printCoreSchema(
+  schema: GraphQLSchema,
+  serviceList: ServiceDefinition[],
+  options?: Options,
+): string {
+  const config = schema.toConfig();
+
+  const {
+    FieldSetScalar,
+    JoinFieldDirective,
+    JoinTypeDirective,
+    JoinOwnerDirective,
+    JoinGraphEnum,
+    JoinGraphDirective,
+    sanitizedServiceNames,
+  } = getJoins(serviceList);
+
+  schema = new GraphQLSchema({
+    ...config,
+    directives: [
+      CoreDirective,
+      JoinFieldDirective,
+      JoinTypeDirective,
+      JoinOwnerDirective,
+      JoinGraphDirective,
+      ...config.directives,
+    ],
+    types: [FieldSetScalar, JoinGraphEnum, ...config.types],
+  });
+
+  const context: PrintingContext = {
+    sanitizedServiceNames,
+  }
+
+  return printFilteredSchema(
+    schema,
+    (n) => !isSpecifiedDirective(n),
+    isDefinedType,
+    options,
+    context,
+  );
+}
+
+export function printIntrospectionSchema(
+  schema: GraphQLSchema,
+  options?: Options,
+): string {
+  return printFilteredSchema(
+    schema,
+    isSpecifiedDirective,
+    isIntrospectionType,
+    options,
+  );
+}
+
+function isDefinedType(type: GraphQLNamedType): boolean {
+  return !isSpecifiedScalarType(type) && !isIntrospectionType(type);
+}
+
+function printFilteredSchema(
+  schema: GraphQLSchema,
+  directiveFilter: (type: GraphQLDirective) => boolean,
+  typeFilter: (type: GraphQLNamedType) => boolean,
+  options?: Options,
+  // Core addition - see `PrintingContext` type for details
+  context?: PrintingContext,
+): string {
+  const directives = schema.getDirectives().filter(directiveFilter);
+  const types = Object.values(schema.getTypeMap())
+    .sort((type1, type2) => type1.name.localeCompare(type2.name))
+    .filter(typeFilter);
+
+  return (
+    [printSchemaDefinition(schema)]
+      .concat(
+        directives.map((directive) => printDirective(directive, options)),
+        types.map((type) => printType(type, options, context)),
+      )
+      .filter(Boolean)
+      .join('\n\n') + '\n'
+  );
+}
+
+function printSchemaDefinition(schema: GraphQLSchema): string {
+  const operationTypes = [];
+
+  const queryType = schema.getQueryType();
+  if (queryType) {
+    operationTypes.push(`  query: ${queryType.name}`);
+  }
+
+  const mutationType = schema.getMutationType();
+  if (mutationType) {
+    operationTypes.push(`  mutation: ${mutationType.name}`);
+  }
+
+  const subscriptionType = schema.getSubscriptionType();
+  if (subscriptionType) {
+    operationTypes.push(`  subscription: ${subscriptionType.name}`);
+  }
+
+  return (
+    'schema' +
+    // Core change: print @core directive usages on schema node
+    printCoreDirectives() +
+    `\n{\n${operationTypes.join('\n')}\n}`
+  );
+}
+
+function printCoreDirectives() {
+  return [
+    'https://lib.apollo.dev/core/v0.1',
+    'https://lib.apollo.dev/join/v0.1',
```

Comment:
Mentioned by @glasser (he'll probably say it in his review too, but just in case): lib -> spec